### DISCUSSION TITLE
From Nodes to Code

### DISCUSSION THREAD

**Discussion by pragma37** [bnpr/Malt#46] [General]
I would love to hear your feedback and questions about the [From Nodes to Code](https://malt3d.com/#/docs/From-Nodes-To-Code/From-Nodes-To-Code) tutorial.

I intend to continue it and would like to know if you find it useful and what would you like to see improved.

**Comment by MeganDrawsaLot** [comment]
One thing I noticed while trying to follow along with the tutorial was that I got a bit confused when I saw the pictures of the nodes.  I thought they would appear in the blender shade editor after I put in the codes, but then I realized you were using them as examples for how the codes worked in a similar way.  It might help to make it clearer that those particular images are not apart of what you see on your screen while following the tutorial.  

Another thing I had trouble with was VSCodes file formats.  I might have missed when it asked about installing them, or I just didn't?  Maybe you could expand on that bit of the tutorial, with some picture examples. 

I'm still trying to figure out glsl so I might have more suggestions as I try out more from the tutorial.  (I know it's short but I've had to take it kind of slow as I try to learn.)  I hope this is helpful so far.

**Comment by pragma37** [comment]
Very helpful. Thank you!

Yep, I see how that can be confusing. I'll update the text to make clearer that the EEVEE nodes are just for reference.

For VSCode, you should see a pop-up at the bottom right asking if you want to install the recommended extensions.
If it didn't pop-up, just search the Microsoft C/C++ extension in the extensions panel (View > Extensions).

I want to record a video-tutorial for all the setup.

**Reply by MeganDrawsaLot** [reply]
Glad I was able too help! This is the first time I've reached out to help develop something in the Blender community - so I'm really happy to hear the positive feedback. Thank you for the info on the extensions as well. :D 
A video tutorial would be awesome!  👍

**Comment by sukumizune** [comment]
Great article, really helpful!

I noticed some things that made noise in my head while following the article.

First, in the Recap code I noticed you initialized the uv_channel variable but didn’t assign a value to it. Is the default 0?

Second, in the Variables code I noticed you wrote:
vec4 semi_transparent_red = vec3(1,0,0,0.5);
Is the assigned vector supposed to match or am I missing something?

The third one I got when writing and now I understand it xD

I’m thankful for your time (:

**Reply by pragma37** [reply]
Thank you!

> First, in the Recap code I noticed you initialized the uv_channel variable but didn’t assign a value to it. Is the default 0?

Yes, uniform values are set to 0 if you don't provide any default value.

> Second, in the Variables code I noticed you wrote:
vec4 semi_transparent_red = vec3(1,0,0,0.5);
Is the assigned vector supposed to match or am I missing something?

Oops! That's a typo. It should be:
```glsl
vec4 semi_transparent_red = vec4(1,0,0,0.5);
```

Edit: fixed
https://malt3d.com/#/docs/From-Nodes-To-Code/From-Nodes-To-Code?id=variables

**Reply by sukumizune** [reply]
Thank you very much! This helped me a lot to get started understanding glsl. I will experiment and read further!

**Comment by HSDSZ** [comment]
Will the formally released MALT be code based? or node based? or both?

**Reply by pragma37** [reply]
Both. Code may be needed for advanced use-cases, though.

**Comment by Kolupsy** [comment]
Does the article still exist, @pragma37? After the docs have been updated I can not find it anymore. It was very useful so it'd be a shame if it wasnt available anymore

**Reply by pragma37** [reply]
It's still in the repository:
https://github.com/bnpr/Malt/blob/Development/docs/From-Nodes-To-Code/From-Nodes-To-Code.md

But it's not on the web since the examples are outdated.
At this point, I think it would make more sense to update it as a tutorial for implementing custom nodes, but that would require an almost complete rewrite.

**Reply by Kolupsy** [reply]
custom nodes as in what I have been doing or are you talking about the nodes generated from GLSL code? I wanted to take a look at how to make a complete shader that can be sourced by a material instead of sourcing from a node tree

**Reply by pragma37** [reply]
> custom nodes as in what I have been doing or are you talking about the nodes generated from GLSL code?

The later.

>  I wanted to take a look at how to make a complete shader that can be sourced by a material instead of sourcing from a node tree

What does sourced by a material mean? 🤔

**Reply by Kolupsy** [reply]
ah haha weird wording. I mean you know how you can generally decide if a material uses a node tree or a file as its source? I meant that the Nodes to Codes tutorial was teaching how to make complete shaders (which is what I wanted to look up again)

**Reply by pragma37** [reply]
Using GLSL files directly is not a great option anymore in the case of the NPR Pipeline, since the Custom IO must be set up from the UI.

Is there any reason you want to follow that workflow?

You could subclass the NPR Pipeline and add the graph IO you need in Python.
Or simply put all your shader code inside one function and use that as a node.

**Reply by Kolupsy** [reply]
> Is there any reason you want to follow that workflow?

Not anymore if you say that this is not a great option anymore 😅

> Using GLSL files directly is not a great option anymore in the case of the NPR Pipeline

does this only apply to Render Nodes or does this apply to the entirety of BlenderMalt? 

The only reason I have for using GLSL files directly is to do multi-pass screen shaders because I dont want to connect multiple Render nodes that are actually part of one 'screen effect'

**Reply by pragma37** [reply]
Oh! I see.

It applies to user-side materials only.
And only if you need user-side custom IO (but in the case of the built-in pipeline that's basically always).

If you are going to compile and run shaders internally from a Python node then plain GLSL is indeed the way to go.
